src/VirtoCommerce.Platform.Core/DistributedLock/DistributedLockResourceBase.cs

```diff
@@ -0,0 +1,77 @@
+using System;
+using RedLockNet.SERedis;
+using RedLockNet.SERedis.Configuration;
+using StackExchange.Redis;
+using VirtoCommerce.Platform.Core.Exceptions;
+
+namespace VirtoCommerce.Platform.Core.DistributedLock
+{
+    /// <summary>
+    /// Basic implementation of distributed lock resource
+    /// </summary>
+    public class DistributedLockResourceBase
+    {
+        /// <summary>
+        /// Resource identifier to use in the distributed lock
+        /// </summary>
+        protected string ResourceId { get; set; } = string.Empty;
+
+        /// <summary>
+        /// Run payload method with distributed lock
+        /// </summary>
+        /// <param name="redisConnMultiplexer">Connection multiplexer pointing to the Redis server, used for locking</param>
+        /// <param name="waitTime">Total time to wait until the lock is available</param>
+        /// <param name="payload">Payload method to run under the acquired lock</param>
+        public virtual void WithLock(IConnectionMultiplexer redisConnMultiplexer, int waitTime, Action<DistributedLockCondition> payload)
+        {
+            if (redisConnMultiplexer != null)
+            {
+                using (var redlockFactory = RedLockFactory.Create(new RedLockMultiplexer[] { new RedLockMultiplexer(redisConnMultiplexer) }))
+                {
+                    var instantlyAcquired = false;
+                    var expiryTime = 120 + waitTime;
+                    // Try to acquire distributed lock giving up immediately if the lock is not available
```

Comment:
// Try to acquire distributed lock and giving up immediately if the lock is not available